src/format-error.ts

```diff
@@ -0,0 +1,47 @@
+import { SourceMapConsumer } from "source-map";
+import * as path from "path";
+
+import type { Bundle } from "./bundle";
+import type { RawSourceMap } from "source-map";
+
+type Formatter = (m: string) => string;
+
+export function createFormatError(bundle: Bundle, formatError?: Formatter) {
+	const consumers = new WeakMap<RawSourceMap, SourceMapConsumer>();
+	const regex = /((?:\b[A-Z]:)?\/[^ #?:]+)[^ :]*:(\d+):(\d+)/gi;
```

Comment:
Parsing stack traces with a single regex is next to impossible due to different stack trace formats in browsers and various tools. The regex here will only match Chrome's format. I ran into that while working on Preact devtools where hooks inspection has to parse the stack trace to find the names of parent closures (=custom hooks).

To make life easier, I collaborated with folks from the [open-wc](https://open-wc.org/) initiative who needed a similar feature for their test runner. They too wanted to improve the cli output like we do in this PR. The result of that can be found in the [errorstacks](https://github.com/marvinhagemeister/errorstacks) package which does all the parsing for you.

Let's replace the regex here with the parser from that package to support all the browser formats 👍

Reply:
I'm a bit hesitant to do that because I this will only give me back the stack frames, and eliminates the error message. Plus, we don't actually need to worry about full frame parsing, only the line/column unminification. The `${url}:${line}:${column}` pattern is consistent across all browsers (though the name and decoration around it are different).

Testing this across browsers, without any sourcemaps, I get:

```
// No sourcemap
Safari 14.0 (Mac OS 10.15.6) simple should work FAILED
	fail
	foo@/var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/4e8f8222d19d100fe2d62c299e035599-bundle.js:4:20
	/var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/4e8f8222d19d100fe2d62c299e035599-bundle.js:10:17
WebKit 537.36 (undefined 0.0.0) simple should work FAILED
	Error: fail
	    at foo (/var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/4e8f8222d19d100fe2d62c299e035599-bundle.js:4:11)
	    at Context.<anonymous> (/var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/4e8f8222d19d100fe2d62c299e035599-bundle.js:10:14)
Firefox 85.0 (Mac OS 10.15) simple should work FAILED
	fail
	foo@/var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/4e8f8222d19d100fe2d62c299e035599-bundle.js:4:11
	@/var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/4e8f8222d19d100fe2d62c299e035599-bundle.js:10:14
```

The old code, where we patched sourcemaps into Karma for it to unminify, generated:
```
// karma sourcemap
Safari 14.0 (Mac OS 10.15.6) simple should work FAILED
	fail
	foo@/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/sub/dep1.js:2:12 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/d7794ccd0c957bf908ce2f6178f29777-bundle.js:4:20
	/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/main-a.js:5:10 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/d7794ccd0c957bf908ce2f6178f29777-bundle.js:10:17
WebKit 537.36 (undefined 0.0.0) simple should work FAILED
	Error: fail
	    at foo (/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/sub/dep1.js:2:8 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/d7794ccd0c957bf908ce2f6178f29777-bundle.js:4:11)
	    at Context.<anonymous> (/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/main-a.js:5:10 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/d7794ccd0c957bf908ce2f6178f29777-bundle.js:10:14)
Firefox 85.0 (Mac OS 10.15) simple should work FAILED
	fail
	foo@/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/sub/dep1.js:2:8 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/d7794ccd0c957bf908ce2f6178f29777-bundle.js:4:11
	@/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/main-a.js:5:10 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/d7794ccd0c957bf908ce2f6178f29777-bundle.js:10:14
```

And, with this code, we get:
```
// formatError
Safari 14.0 (Mac OS 10.15.6) simple should work FAILED
	fail
	foo@/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/sub/dep1.js:2:12 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/634d590ecfca5cf2f74e2f551503479b-bundle.js:4:20
	/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/main-a.js:5:10 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/634d590ecfca5cf2f74e2f551503479b-bundle.js:10:17
WebKit 537.36 (undefined 0.0.0) simple should work FAILED
	Error: fail
	    at foo (/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/sub/dep1.js:2:8 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/634d590ecfca5cf2f74e2f551503479b-bundle.js:4:11)
	    at Context.<anonymous> (/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/main-a.js:5:10 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/634d590ecfca5cf2f74e2f551503479b-bundle.js:10:14)
Firefox 85.0 (Mac OS 10.15) simple should work FAILED
	fail
	foo@/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/sub/dep1.js:2:8 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/634d590ecfca5cf2f74e2f551503479b-bundle.js:4:11
	@/Users/jridgewell/src/karma-esbuild/test/fixtures/sourcemap-error-progress/files/main-a.js:5:10 <- /var/folders/0q/yfkhh_05215drnbzwyy9x6zr0000gq/T/634d590ecfca5cf2f74e2f551503479b-bundle.js:10:14
```

For reference, the Karma sourcemap unminification is very similar: https://github.com/karma-runner/karma/blob/d3ff91a5a2f7ff84141096bb47aa31ff73f81cca/lib/reporter.js#L51-L83

Reply:
Thanks for verifying it with the current state of browsers. We can always revisit it later again when a parsing issue pops up. But since it works with current browsers I'm good 👍